From 417ad84bec5d4f9c64e302669f82c6a84b0e8629 Mon Sep 17 00:00:00 2001
From: "GONG, Ruiqi" <gongruiqi1@huawei.com>
Date: Thu, 20 Jan 2022 19:06:56 +0800
Subject: [PATCH 1491/2065] [Huawei] selinux: genfscon ns: Support multiple
 isec for an inode

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186697

----------------------------------------

Since one isec corresponds to one label, for an inode that supports
genfscon namespace, processes in different namespace should retrieve
different isec from the inode. So change i_security to a list, and
before processes can create and switch to new namespace, always access
the first entry in the list, which corresponds to the init namespace
and is the only available one for now.

Signed-off-by: GONG, Ruiqi <gongruiqi1@huawei.com>
---
 security/selinux/hooks.c            | 117 ++++++++++++++++++++--------
 security/selinux/include/objsec.h   |  11 ++-
 security/selinux/include/security.h |   2 +
 security/selinux/selinuxfs.c        |   4 +-
 4 files changed, 99 insertions(+), 35 deletions(-)

diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 51a983439cb7..72a3b5239654 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -111,6 +111,7 @@ static atomic_t selinux_secmark_refcount = ATOMIC_INIT(0);
 
 const char *SELINUX_NS_INIT_NAME = "";
 static struct selinux_ns init_selinux_ns;
+struct kmem_cache *inode_security_cache;
 
 #ifdef CONFIG_SECURITY_SELINUX_DEVELOP
 static int selinux_enforcing_boot __initdata;
@@ -193,6 +194,26 @@ static int __init selinux_ns_fs_setup(char *str)
 }
 __setup("selinux_ns_fs=", selinux_ns_fs_setup);
 
+static struct inode_security_struct *isec_alloc(struct inode *inode)
+{
+	struct inode_security_struct *isec = kmem_cache_zalloc(
+			inode_security_cache, GFP_NOFS);
+	if (!isec)
+		return NULL;
+
+	spin_lock_init(&isec->lock);
+	INIT_LIST_HEAD(&isec->list);
+	INIT_LIST_HEAD(&isec->isec_list);
+	isec->sid = SECINITSID_UNLABELED;
+	isec->sclass = SECCLASS_FILE;
+	isec->task_sid = current_sid();
+	isec->initialized = LABEL_INVALID;
+	isec->inode = inode;
+	isec->ns = &init_selinux_ns;
+
+	return isec;
+}
+
 /**
  * selinux_secmark_enabled - Check to see if SECMARK is currently enabled
  *
@@ -286,6 +307,12 @@ static inline u32 task_sid(const struct task_struct *task)
 
 static int inode_doinit_with_dentry(struct inode *inode, struct dentry *opt_dentry);
 
+struct inode_security_struct *inode_security_novalidate(struct inode *inode)
+{
+	return container_of(selinux_inode(inode)->head.next,
+			struct inode_security_struct, isec_list);
+}
+
 /*
  * Try reloading inode security labels that have been marked as invalid.  The
  * @may_sleep parameter indicates when sleeping and thus reloading labels is
@@ -296,7 +323,7 @@ static int __inode_security_revalidate(struct inode *inode,
 				       struct dentry *dentry,
 				       bool may_sleep)
 {
-	struct inode_security_struct *isec = selinux_inode(inode);
+	struct inode_security_struct *isec = inode_security_novalidate(inode);
 
 	might_sleep_if(may_sleep);
 
@@ -315,11 +342,6 @@ static int __inode_security_revalidate(struct inode *inode,
 	return 0;
 }
 
-static struct inode_security_struct *inode_security_novalidate(struct inode *inode)
-{
-	return selinux_inode(inode);
-}
-
 static struct inode_security_struct *inode_security_rcu(struct inode *inode, bool rcu)
 {
 	int error;
@@ -327,7 +349,7 @@ static struct inode_security_struct *inode_security_rcu(struct inode *inode, boo
 	error = __inode_security_revalidate(inode, NULL, !rcu);
 	if (error)
 		return ERR_PTR(error);
-	return selinux_inode(inode);
+	return inode_security_novalidate(inode);
 }
 
 /*
@@ -336,14 +358,14 @@ static struct inode_security_struct *inode_security_rcu(struct inode *inode, boo
 static struct inode_security_struct *inode_security(struct inode *inode)
 {
 	__inode_security_revalidate(inode, NULL, true);
-	return selinux_inode(inode);
+	return inode_security_novalidate(inode);
 }
 
 static struct inode_security_struct *backing_inode_security_novalidate(struct dentry *dentry)
 {
 	struct inode *inode = d_backing_inode(dentry);
 
-	return selinux_inode(inode);
+	return inode_security_novalidate(inode);
 }
 
 /*
@@ -354,15 +376,27 @@ static struct inode_security_struct *backing_inode_security(struct dentry *dentr
 	struct inode *inode = d_backing_inode(dentry);
 
 	__inode_security_revalidate(inode, dentry, true);
-	return selinux_inode(inode);
+	return inode_security_novalidate(inode);
+}
+
+static void isech_free_rcu(struct rcu_head *head)
+{
+	struct inode_security_head *isech;
+	struct inode_security_struct *entry, *tmp;
+
+	isech = container_of(head, struct inode_security_head, rcu);
+	list_for_each_entry_safe(entry, tmp, &isech->head, isec_list) {
+		kmem_cache_free(inode_security_cache, entry);
+	}
 }
 
 static void inode_free_security(struct inode *inode)
 {
-	struct inode_security_struct *isec = selinux_inode(inode);
+	struct inode_security_head *isech = selinux_inode(inode);
+	struct inode_security_struct *isec;
 	struct superblock_security_struct *sbsec;
 
-	if (!isec)
+	if (!isech)
 		return;
 	sbsec = inode->i_sb->s_security;
 	/*
@@ -375,11 +409,24 @@ static void inode_free_security(struct inode *inode)
 	 * concurrent list_add(), but for better safety against future changes
 	 * in the code, we use list_empty_careful() here.
 	 */
-	if (!list_empty_careful(&isec->list)) {
-		spin_lock(&sbsec->isec_lock);
-		list_del_init(&isec->list);
-		spin_unlock(&sbsec->isec_lock);
+	list_for_each_entry(isec, &isech->head, isec_list) {
+		if (!list_empty_careful(&isec->list)) {
+			spin_lock(&sbsec->isec_lock);
+			list_del_init(&isec->list);
+			spin_unlock(&sbsec->isec_lock);
+		}
 	}
+
+	/*
+	 * The inode may still be referenced in a path walk and
+	 * a call to selinux_inode_permission() can be made
+	 * after inode_free_security() is called. Ideally, the VFS
+	 * wouldn't do this, but fixing that is a much harder
+	 * job. For now, simply free the i_security via RCU, and
+	 * leave the current inode->i_security pointer intact.
+	 * The inode will be freed after the RCU grace period too.
+	 */
+	call_rcu(&isech->rcu, isech_free_rcu);
 }
 
 static void superblock_free_security(struct super_block *sb)
@@ -1433,7 +1480,7 @@ static int inode_doinit_use_xattr(struct inode *inode, struct dentry *dentry,
 static int inode_doinit_with_dentry(struct inode *inode, struct dentry *opt_dentry)
 {
 	struct superblock_security_struct *sbsec = NULL;
-	struct inode_security_struct *isec = selinux_inode(inode);
+	struct inode_security_struct *isec = inode_security_novalidate(inode);
 	u32 task_sid, sid = 0;
 	u16 sclass;
 	struct dentry *dentry;
@@ -1691,7 +1738,7 @@ static int inode_has_perm(const struct cred *cred,
 		return 0;
 
 	sid = cred_sid(cred);
-	isec = selinux_inode(inode);
+	isec = inode_security_novalidate(inode);
 
 	return avc_has_perm(&selinux_state,
 			    sid, isec->sid, isec->sclass, perms, adp);
@@ -2873,16 +2920,17 @@ static int selinux_fs_context_parse_param(struct fs_context *fc,
 
 static int selinux_inode_alloc_security(struct inode *inode)
 {
-	struct inode_security_struct *isec = selinux_inode(inode);
-	u32 sid = current_sid();
+	struct inode_security_struct *isec;
+	struct inode_security_head *isech = selinux_inode(inode);
 
-	spin_lock_init(&isec->lock);
-	INIT_LIST_HEAD(&isec->list);
-	isec->inode = inode;
-	isec->sid = SECINITSID_UNLABELED;
-	isec->sclass = SECCLASS_FILE;
-	isec->task_sid = sid;
-	isec->initialized = LABEL_INVALID;
+	isec = isec_alloc(inode);
+	if (!isec)
+		return -ENOMEM;
+
+	spin_lock_init(&isech->lock);
+	INIT_LIST_HEAD(&isech->head);
+	isech->task_sid = current_sid();
+	list_add(&isec->isec_list, &isech->head);
 
 	return 0;
 }
@@ -2954,7 +3002,9 @@ static int selinux_inode_init_security(struct inode *inode, struct inode *dir,
 
 	/* Possibly defer initialization to selinux_complete_init. */
 	if (sbsec->flags & SE_SBINITIALIZED) {
-		struct inode_security_struct *isec = selinux_inode(inode);
+		struct inode_security_struct *isec;
+
+		isec = inode_security_novalidate(inode);
 		isec->sclass = inode_mode_to_security_class(inode->i_mode);
 		isec->sid = newsid;
 		isec->initialized = LABEL_INITIALIZED;
@@ -3054,7 +3104,7 @@ static noinline int audit_inode_permission(struct inode *inode,
 					   int result)
 {
 	struct common_audit_data ad;
-	struct inode_security_struct *isec = selinux_inode(inode);
+	struct inode_security_struct *isec = inode_security_novalidate(inode);
 	int rc;
 
 	ad.type = LSM_AUDIT_DATA_INODE;
@@ -4214,7 +4264,7 @@ static int selinux_task_kill(struct task_struct *p, struct kernel_siginfo *info,
 static void selinux_task_to_inode(struct task_struct *p,
 				  struct inode *inode)
 {
-	struct inode_security_struct *isec = selinux_inode(inode);
+	struct inode_security_struct *isec = inode_security_novalidate(inode);
 	u32 sid = task_sid(p);
 
 	spin_lock(&isec->lock);
@@ -6551,7 +6601,7 @@ static void selinux_release_secctx(char *secdata, u32 seclen)
 
 static void selinux_inode_invalidate_secctx(struct inode *inode)
 {
-	struct inode_security_struct *isec = selinux_inode(inode);
+	struct inode_security_struct *isec = inode_security_novalidate(inode);
 
 	spin_lock(&isec->lock);
 	isec->initialized = LABEL_INVALID;
@@ -6927,7 +6977,7 @@ static int selinux_lockdown(enum lockdown_reason what)
 struct lsm_blob_sizes selinux_blob_sizes __lsm_ro_after_init = {
 	.lbs_cred = sizeof(struct task_security_struct),
 	.lbs_file = sizeof(struct file_security_struct),
-	.lbs_inode = sizeof(struct inode_security_struct),
+	.lbs_inode = sizeof(struct inode_security_head),
 	.lbs_ipc = sizeof(struct ipc_security_struct),
 	.lbs_msg_msg = sizeof(struct msg_security_struct),
 };
@@ -7299,6 +7349,9 @@ static __init int selinux_init(void)
 
 	default_noexec = !(VM_DATA_DEFAULT_FLAGS & VM_EXEC);
 
+	inode_security_cache = kmem_cache_create("selinux_inode_security",
+					    sizeof(struct inode_security_struct),
+					    0, SLAB_PANIC, NULL);
 	avc_init();
 
 	avtab_cache_init();
diff --git a/security/selinux/include/objsec.h b/security/selinux/include/objsec.h
index c3bbfc8d3bc4..aa84b7d833b4 100644
--- a/security/selinux/include/objsec.h
+++ b/security/selinux/include/objsec.h
@@ -44,6 +44,13 @@ enum label_initialized {
 	LABEL_PENDING
 };
 
+struct inode_security_head {
+	struct list_head head;	/* list head of inode_security_struct */
+	u32 task_sid;		/* the real SID of inode creater */
+	spinlock_t lock;
+	struct rcu_head rcu;	/* for freeing the inode_security_struct */
+};
+
 struct inode_security_struct {
 	struct inode *inode;	/* back pointer to inode object */
 	struct list_head list;	/* list of inode_security_struct */
@@ -52,6 +59,8 @@ struct inode_security_struct {
 	u16 sclass;		/* security class of this object */
 	unsigned char initialized;	/* initialization flag */
 	spinlock_t lock;
+	struct list_head isec_list;	/* list of isec */
+	struct selinux_ns *ns;	/* corresponding namespace */
 };
 
 struct file_security_struct {
@@ -160,7 +169,7 @@ static inline struct file_security_struct *selinux_file(const struct file *file)
 	return file->f_security + selinux_blob_sizes.lbs_file;
 }
 
-static inline struct inode_security_struct *selinux_inode(
+static inline struct inode_security_head *selinux_inode(
 						const struct inode *inode)
 {
 	if (unlikely(!inode->i_security))
diff --git a/security/selinux/include/security.h b/security/selinux/include/security.h
index 25d0a02ed42a..63e5c31f0922 100644
--- a/security/selinux/include/security.h
+++ b/security/selinux/include/security.h
@@ -460,5 +460,7 @@ extern void avtab_cache_init(void);
 extern void ebitmap_cache_init(void);
 extern void hashtab_cache_init(void);
 extern int security_sidtab_hash_stats(struct selinux_state *state, char *page);
+extern struct inode_security_struct *inode_security_novalidate(
+		struct inode *inode);
 
 #endif /* _SELINUX_SECURITY_H_ */
diff --git a/security/selinux/selinuxfs.c b/security/selinux/selinuxfs.c
index 2b745ae8cb98..09df13a656df 100644
--- a/security/selinux/selinuxfs.c
+++ b/security/selinux/selinuxfs.c
@@ -1458,7 +1458,7 @@ static int sel_make_bools(struct selinux_policy *newpolicy, struct dentry *bool_
 			goto out;
 		}
 
-		isec = selinux_inode(inode);
+		isec = inode_security_novalidate(inode);
 		ret = selinux_policy_genfs_sid(newpolicy, "selinuxfs", page,
 					 SECCLASS_FILE, &sid);
 		if (ret) {
@@ -2109,7 +2109,7 @@ static int sel_fill_super(struct super_block *sb, struct fs_context *fc)
 	}
 
 	inode->i_ino = ++fsi->last_ino;
-	isec = selinux_inode(inode);
+	isec = inode_security_novalidate(inode);
 	isec->sid = SECINITSID_DEVNULL;
 	isec->sclass = SECCLASS_CHR_FILE;
 	isec->initialized = LABEL_INITIALIZED;
-- 
2.17.1

